### DISCUSSION TITLE
Update prompt regularly

### DISCUSSION THREAD

**Discussion by killermoehre** [akinomyoga/ble.sh#341] [Q&A] [answered]
Hi,

the terminal emulator of my choice (`xfce4-terminal`, `libvte` based) doesn't support a status line. But I would really like to see the current time in the lower right corner (and on a `tty` and such).

I can use

```bash
bleopt prompt_status_line='\r\r\t'
bleopt prompt_status_align=$'justify=\r'
```

to show the time stamp, but it's not refreshing every second.

Any way to implement such feature? Or would I need to use some additional tooling like `tmux` (which would be hard overkill for this, IMHO)?

**Comment by akinomyoga** [comment, marked as answer]
This is an example (You need Bash 4.0+):

```bash
# blerc

ble/util/idle.push 'ble/util/idle.sleep 1000'
function ble/prompt/backslash:killermoehre/time {
  ble/prompt/unit/add-hash '$SECONDS'
  ble/prompt/process-prompt-string '\t'
}
bleopt prompt_status_line='\r\q{killermoehre/time}'
bleopt prompt_status_align=$'justify=\r'
```

`ble/prompt/unit/add-hash '$SECONDS'` instructs that the prompt content should be invalidated when the Bash built-in shell variable `SECONDS` is changed. `ble/util/idle.push 'ble/util/idle.sleep 1000'` is an empty timer loop to induce an update if any of the UI elements is invalidated.

**Reply by killermoehre** [reply]
You know you're aweseome ♥️

**Reply by killermoehre** [reply]
Question:

Can the status prompt kept visible *while other applications run*? Right now, the status line goes away as soon as another application/program (i.e. `sleep 10`) run.

**Reply by akinomyoga** [reply]
I don't think there is a robust way to achieve that without the support of the terminal. There are two problems.
- When the user command is running, the terminal display is supposed to be freely changed by the user command, i.e., even if `ble.sh` doesn't erase the status prompt before starting the user command, the user command can overwrite the status line.
- Also, while an external command is running, Bash waits for the termination of the external command, i.e., it stops working until the external command completes. So, even if there is a way to protect the status line from the user command, there is no way for `ble.sh` to update the status line in the middle of the external command execution.

**Reply by killermoehre** [reply]
I see, I see. Well, even then your solution is marvellous. Thank you very much 😊

**Reply by killermoehre** [reply]
@akinomyoga I used your solution now. Still great. But could be greater :sweat_smile: 

Do you know how I could configure `tmux` in such a way that it *only* renders the status bar, but does noting else? Doesn't take over the alternative screen, works with scroll wheel input as expected, etc.

**Reply by killermoehre** [reply]
I just found [shox](https://github.com/liamg/shox), which might do what I want.

**Reply by killermoehre** [reply]
Nope, `shox` is not what I want.

**Reply by akinomyoga** [reply]
> Do you know how I could configure `tmux` in such a way that it _only_ renders the status bar, but does noting else?

I'm not sure, but it depends on what you expect for the detailed behaviors.

> Doesn't take over the alternative screen,

I'm not sure if I correctly understand what you mean, but maybe you can check the following Stack Overflow answer:

- [macos - Why does tmux erase terminal contents on editor exit? - Stack Overflow](https://stackoverflow.com/questions/4292572/why-does-tmux-erase-terminal-contents-on-editor-exit)

In your case, you might need to turn off the alternate screen as
```
# .tmux.conf

set-window-option -g alternate-screen off
```

> works with scroll wheel input as expected, etc.

Maybe you could check the following page. You may use `set -g mouse on` or `set -g mouse off` depending on whether mouse scroll is processed inside/outside `tmux`.

- [Scroll shell output with mouse in tmux - Super User](https://superuser.com/questions/210125/scroll-shell-output-with-mouse-in-tmux)

**Reply by akinomyoga** [reply]
Today, there was a related post for a status line in the `help-bash` mailing list, though it doesn't seem to be so robust as far as I tested it

https://lists.gnu.org/archive/html/help-bash/2023-08/msg00002.html
